### ISSUE TITLE
Can ZoKrates proofs bundle be hosted remotely?

### ISSUE THREAD

**Issue by 1blockologist** [EYBlockchain/nightfall#4]
Can the 7GB of proofs be hosted on a remote system to allow for lighter clients and systems with less resources to use this system?

Basically a trusted setup but where the result is that the onchain transactions are still just as secure and obfuscated.

**Comment by Chaitanya-Konda** [comment]
The proof files are over 7GB and GitHub doesn't accept files that big to add to this repo. Also, another major risk is that, providing proofs require that their proving keys and verification keys to be provided as well. Which could be misused by people in their applications. 

On the other hand, the README.md explains how to generate them on your local machine.

**Comment by 1blockologist** [comment]
I'm asking about the ability to able to generate them on own computer and use them on my mobile phone to create private transactions.

What are the challenges to that?

There are many situations where 16GB of RAM, and 10GB of storage space are not practical for nodes

**Comment by Chaitanya-Konda** [comment]
`Installing Nightfall` section under README.md does exactly what you are asking for. To generate them on your own computer. ZKPs as a technology are yet to become practical to be generated on a mobile

**Comment by 1blockologist** [comment]
Can the ZKPs that I generate on my computer be used by a transaction signing process initiated by a mobile device, where the mobile device queries ZKPs remotely, or sends a raw transaction to my computer and somehow factors in the ZKPs to that raw transaction before broadcasting it to the Ethereum blockchain?

**Comment by Westlad** [comment]
Yes in the sense that you could compute a proof remotely (and would probably want to given the heaviness of the compute) but you'd have to trust the hardware it was running on though.

**Comment by 1blockologist** [comment]
@Westlad yes, there are different levels of trust needed depending on the application. I would factor in whether my private key can be derived, or the transaction could be commandeered to reappropriate actual funds to another party, or whether someone else can read that a transaction is happening, or whether I only care about what can be seen on the blockchain itself

I'm curious what possibilities there are

**Comment by Genysys** [comment]
An implementation sonics might be worth exploring. Doesnt use a trusted set
up and proofs that are constant and can be linear which makes signature
aggregation feasible.

https://eprint.iacr.org/2019/099.pdf

**Comment by vadym-f** [comment]
> .. Also, another major risk is that, providing proofs require that their proving keys and verification keys to be provided as well. Which could be misused by people in their applications.

Would you mind explaining exactly what is expected risk of exposing proofs and keys please?
Zcash proofs and public keys are available/open to the world, still alive and doing well.

**Comment by Chaitanya-Konda** [comment]
@vadym-f In order to generate proving and verifying keys, a trusted setup needs to be done. Trusted setup takes certain parameters(randomness) along with the circuit to generate the keys. If whoever performs the trusted setup decides to not discard the parameters, then they will be able to use these to generate fake proofs that will satisfy the verification. Which means counterfeit tokens can be created. 

Zcash handles does the trusted setup by doing a Multi Party Computation ceremony where a large number of independent parties come together to generate these parameters. All of the participants would have to be compromised or dishonest in order to compromise the final parameters. 

Doing an MPC ceremony of that extent to provide keys that can be used to test nightfall source code is unrealistic. And if no MPC ceremony is done and keys are provided as part of the source code by nightfall team, then anyone who forks nightfall and use it in their real world application might not realise to use a different set of keys. In which case, if the the team of nightfall has access to the parameters used to generate these keys still, they would be able to create fake tokens of whatever real world application that is. 

In short, it is not best practise in terms of security. 
Similar in line of reasoning to, putting private keys (public key encryption) or private keys of an Ethereum address in GitHub source code repositories being silly.

**Comment by ChaitanyaKonda** [comment]
> An implementation sonics might be worth exploring. Doesnt use a trusted set up and proofs that are constant and can be linear which makes signature aggregation feasible. https://eprint.iacr.org/2019/099.pdf
> […](#)

@Genysys We work very closely with the author of Sonics, Mary Maller. We are considering to use Sonics.
